src/util/FeedSwitch.sol

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.18;
+
+import "src/interfaces/IChainlinkFeed.sol";
+
+interface IPendlePT {
+    function expiry() external view returns (uint256);
+}
+
+/// @title FeedSwitch for Pendle PT tokens
+/// @notice A contract to switch between feeds after a timelock period
+/// @dev The switch can only be initiated by the guardian and will be effective after the timelock period has passed, can only be done before maturity.
+/// If the switch is initiated but not yet effective, it can be cancelled by the guardian without timelock period.
+/// The feed will default to the afterMaturityFeed if the maturity has passed.
+/// The guardian can initiate the switch again to the previous feed multiple times.
+contract FeedSwitch {
+    error NotGuardian();
+    error FeedDecimalsMismatch();
+    error MaturityInPast();
+    error MaturityPassed();
+
+    address public immutable guardian;
+    IChainlinkFeed public immutable initialFeed;
+    IChainlinkFeed public immutable beforeMaturityFeed;
+    IChainlinkFeed public immutable afterMaturityFeed;
+    uint256 public immutable maturity;
+    uint256 public immutable timelockPeriod;
+
+    IChainlinkFeed public feed;
+    IChainlinkFeed public previousFeed;
+
+    uint256 public switchCompletedAt;
+
+    event FeedSwitchInitiated(address indexed newFeed);
+
+    constructor(
+        address _feed,
+        address _beforeMaturityFeed,
+        address _afterMaturityFeed,
+        uint256 _timelockPeriod,
+        address _pendlePT,
+        address _guardian
+    ) {
+        feed = IChainlinkFeed(_feed);
+        initialFeed = IChainlinkFeed(_feed);
+        beforeMaturityFeed = IChainlinkFeed(_beforeMaturityFeed);
+        afterMaturityFeed = IChainlinkFeed(_afterMaturityFeed);
+        if (
+            beforeMaturityFeed.decimals() != 18 ||
+            afterMaturityFeed.decimals() != 18 ||
+            feed.decimals() != 18
+        ) revert FeedDecimalsMismatch();
+
+        timelockPeriod = _timelockPeriod;
+        maturity = IPendlePT(_pendlePT).expiry();
+        if (maturity < block.timestamp) revert MaturityInPast();
+
+        guardian = _guardian;
+    }
+
+    /// @notice Initiate the feed switch, entering the timelock period
+    /// @dev Can only be called by the guardian and can be done before maturity, after that, the feed will return afterMaturityFeed data
+    function initiateFeedSwitch() external {
+        if (msg.sender != guardian) revert NotGuardian();
+        if (block.timestamp > maturity) revert MaturityPassed();
+
+        if (switchCompletedAt < block.timestamp) {
```

Comment:
suggestion:
from UX pov, not great to have a single 0-parameter function do multiple different things.
1. queue switch
1. cancel switch

could we benefit from a helper funcs like this?:
- `isFeedSwitchQueued() public returns (bool)`

... may not be strictly necessary, but could help guardian know exactly what the consequence of their next call will be.